Add Everest Photosensitivity menu

GamationOnGithub opened the pull request:
Adds a submenu to the vanilla Options menu that allows for toggling the 5 parts of Photosensitive Mode (anxiety, glitch effects, screen flashes, lightning flashes, and text highlighting) on and off. Also creates publicly accessibly options for modders to tie their own effects to, e.g. `AllowFlash`, `AllowTextHighlighting`, etc.

GamationOnGithub committed:
Add photosensitivity options to mod menu

GamationOnGithub committed:
Adjust property naming

Rename split photosensitivity settings, add public-facing properties

GamationOnGithub committed:
Add functionality for Glitch suppression

GamationOnGithub committed:
Add functionality for Distort suppression

No IL patch here thank god

GamationOnGithub committed:
Add support for suppressing screen flashes

Also remove an extraneous comment

GamationOnGithub committed:
Text Highlighting, pt 1 (the easy stuff)

GamationOnGithub committed:
Text Highlighting pt 2: IL patches

Only coroutines remain, god save my wretched soul. Also fix an unnecessary "new" declaration

GamationOnGithub committed:
i am so good at version management software

GamationOnGithub committed:
Revert "i am so good at version management software"

This reverts commit a1944f5b91014fc52f9f9ea3a518038229490e8c.

GamationOnGithub committed:
Finish text highlighting support

GamationOnGithub committed:
Add Lightning support

fuck youuuu coroutine IL patches

GamationOnGithub committed:
Move menu from Mod Options to vanilla Options

GamationOnGithub committed:
Dialog fixes

GamationOnGithub committed:
Give menu items descriptions

GamationOnGithub commented:
Notably:
- I left in the Mod Options implementation in case we decide it's better there instead of vanilla options, it's just commented out. should probably remove that at some point if we commit to going with vanilla options
- Adding the descriptions is kind of redundant as all fuck and can definitely be made better

GamationOnGithub committed:
Remove pragma balls

Sleep tight sweet prince

Kalobi commented on Celeste.Mod.mm/Patches/LightningRenderer.cs:
```diff
@@ -0,0 +1,103 @@
+namespace Celeste {
+    class patch_LightningRenderer : LightningRenderer {
+
+        [MonoModPublic]
```
MonoModPublic makes the class and methods public even in the patched executable, which can break reflection.

GamationOnGithub replied:
Resolved in commit [6f4ef7f](https://github.com/EverestAPI/Everest/pull/836/commits/6f4ef7fd93c9b4f858749dc54c82c2e1e52a41e8)

Kalobi commented on Celeste.Mod.mm/Mod/Core/CoreModuleSettings.cs:
```diff
@@ -308,6 +308,56 @@ public bool ColorizedLogging {
+        }
+        
+        [SettingIgnore]
+        public bool PhotosensitivityDistortOverride { get; set; } = false;
```
I still think having two public properties for each setting is very confusing. Ideally we should figure out a way to implement this while only exposing the Allow... properties (I'm convinced this is possible, YamlDotNet is just poorly documented unfortunately). Barring that, we should at least indicate that these shouldn't be used with something like the Obsolete attribute (though that doesn't quite fit).

GamationOnGithub replied:
Addressed in commit [60e0aed](https://github.com/EverestAPI/Everest/pull/836/commits/60e0aed887428e74f89386f01090e3982a857437) using implementation proposed in [#modding-dev](https://discord.com/channels/403698615446536203/429775439423209472/1300537022900867094)

Kalobi commented on Celeste.Mod.mm/Mod/Core/CoreModuleSettings.cs:
```diff
@@ -308,6 +308,56 @@ public bool ColorizedLogging {
+
+        [SettingIgnore]
+        [YamlIgnore]
+        public bool AllowDistort {
+            get => !Settings.Instance.DisableFlashes || PhotosensitivityDistortOverride; 
+        }
```
```suggestion
        public bool AllowDistort => !Settings.Instance.DisableFlashes || PhotosensitivityDistortOverride; 
```
Same applies in a bunch of places.

GamationOnGithub replied:
Resolved in commit [6f4ef7f](https://github.com/EverestAPI/Everest/pull/836/commits/6f4ef7fd93c9b4f858749dc54c82c2e1e52a41e8)

GamationOnGithub committed:
Code cleanup

No longer breaks reflection and makes getters more efficient

GamationOnGithub committed:
Adequately document new photosensitivity options

maddie480 commented:
/azp run

Kalobi commented on Celeste.Mod.mm/Patches/Distort.cs:
```diff
@@ -0,0 +1,22 @@
+            set 
+            {
+                anxiety = value;
+                GFX.FxDistort.Parameters["anxiety"].SetValue((CoreModule.Settings.AllowDistort) ? anxiety : 0f);
```
```suggestion
                GFX.FxDistort.Parameters["anxiety"].SetValue(CoreModule.Settings.AllowDistort ? anxiety : 0f);
```

Kalobi committed:
Remove unnecessary parentheses

SnipUndercover commented on Celeste.Mod.mm/Patches/Glitch.cs:
```diff
@@ -0,0 +1,47 @@
+            cursor.GotoNext(MoveType.Before, instr => instr.MatchLdsfld("Celeste.Settings", "Instance"));
+            cursor.RemoveRange(2);
+            cursor.EmitCall(m_CoreModule_get_Settings);
+            cursor.EmitCall(m_get_AllowGlitch);
```
```suggestion
            cursor.EmitCallvirt(m_get_AllowGlitch);
```

Wouldn't `EmitCallvirt` be more appropriate here? Since `AllowGlitch` is an instance property, the compiler would emit a `callvirt` instead of a `call`.

GamationOnGithub replied:
Addressed in commit [a87a04f](https://github.com/EverestAPI/Everest/pull/836/commits/a87a04fdafc78728fdb24bff92cbe487f25b805b)

SnipUndercover commented on Celeste.Mod.mm/Patches/LightningRenderer.cs:
```diff
@@ -0,0 +1,101 @@
+
+                // Get the AllowLightning value 
+                cursor.Index++;
+                cursor.Next.OpCode = OpCodes.Call;
```
```suggestion
                cursor.Next.OpCode = OpCodes.Callvirt;
```

Same here - `AllowLightning` is an instance property.

GamationOnGithub replied:
Addressed in commit [a87a04f](https://github.com/EverestAPI/Everest/pull/836/commits/a87a04fdafc78728fdb24bff92cbe487f25b805b)

SnipUndercover commented on Celeste.Mod.mm/Patches/LightningRenderer.cs:
```diff
@@ -0,0 +1,101 @@
+
+            // Get the AllowLightning value 
+            cursor.Index++;
+            cursor.Next.OpCode = OpCodes.Call;
```
```suggestion
            cursor.Next.OpCode = OpCodes.Callvirt;
```

Same here - `AllowLightning` is an instance property.

GamationOnGithub replied:
Addressed in commit [a87a04f](https://github.com/EverestAPI/Everest/pull/836/commits/a87a04fdafc78728fdb24bff92cbe487f25b805b)

SnipUndercover commented on Celeste.Mod.mm/Patches/OuiFileSelectSlot.cs:
```diff
@@ -338,6 +356,20 @@ class PatchOuiFileSelectSlotOnContinueSelectedAttribute : Attribute { }
+            cursor.GotoNext(MoveType.Before, instr => instr.MatchLdsfld("Celeste.Settings", "Instance"));
+            cursor.RemoveRange(2);
+            cursor.EmitCall(m_CoreModule_get_Settings);
+            cursor.EmitCall(m_get_AllowScreenFlash);
```
```suggestion
            cursor.EmitCallvirt(m_get_AllowScreenFlash);
```

Same here - `AllowScreenFlash` is an instance property.

GamationOnGithub replied:
Addressed in commit [a87a04f](https://github.com/EverestAPI/Everest/pull/836/commits/a87a04fdafc78728fdb24bff92cbe487f25b805b)

SnipUndercover commented on Celeste.Mod.mm/Patches/OuiMainMenu.cs:
```diff
@@ -9,6 +9,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Runtime.CompilerServices;
```
```suggestion
```
Looks like an accidental using.

GamationOnGithub replied:
Addressed in commit [a87a04f](https://github.com/EverestAPI/Everest/pull/836/commits/a87a04fdafc78728fdb24bff92cbe487f25b805b)

SnipUndercover commented on Celeste.Mod.mm/Patches/TextMenu.cs:
```diff
@@ -587,6 +594,31 @@ public static void PatchTextMenuOptionColor(ILContext context, CustomAttribute a
+
+            // Get the AllowTextHighlight value 
+            cursor.Index++;
+            cursor.Next.OpCode = OpCodes.Call;
```
```suggestion
            cursor.Next.OpCode = OpCodes.Callvirt;
```

Same here - `AllowTextHighlight` is an instance property.

GamationOnGithub replied:
Addressed in commit [a87a04f](https://github.com/EverestAPI/Everest/pull/836/commits/a87a04fdafc78728fdb24bff92cbe487f25b805b)

SnipUndercover reviewed (approved):
Gave my two cents. LGTM, though I think some `call`s would be more accurate as `callvirt`s if we're going by what the compiler would emit.

GamationOnGithub committed:
IL Cleanup

Change some `call`s to `callvirt`s to more accurately match non-artificial IL. Also remove an unnecessary `using`
